### ISSUE TITLE
reverse_bytecode_order problem

### ISSUE THREAD

**Issue by sebseb7** [michaelkamprath/bespokeasm#16]
it works for opcode 0x24 but not for opcode 0x25 (partially)

register dd is 0xa0, for opcode 0x24 it is placed last, for 0x25 placed first.

https://github.com/sebseb7/logicpu/blob/master/isa.yaml

```
      34 |     ld dd,0x81,[var4]                  |  0x8030 | 0x25 | 00100101 | 
         |                                        |  0x8031 | 0xa0 | 10100000 |
         |                                        |  0x8032 | 0x00 | 00000000 |
         |                                        |  0x8033 | 0x02 | 00000010 |
         |                                        |  0x8034 | 0x81 | 10000001 |
      35 |     ld dd,b,c                          |  0x8035 | 0x24 | 00100100 | 
         |                                        |  0x8036 | 0x32 | 00110010 |
         |                                        |  0x8037 | 0xa0 | 10100000
```
```
      - var4b:
        byte_code:
          value: 36
          size: 8
        operands:
          count: 3
          operand_sets:
            reverse_bytecode_order: true
            list:
              - wo_register
              - register
              - register
      - var6:
        byte_code:
          value: 37
          size: 8
        operands:
          count: 3
          operand_sets:
            reverse_bytecode_order: true
            list:
              - register_except_a
              - integer
              - address
```

**Comment by michaelkamprath** [comment]
In the `ld` variant your named `var6` (not that field is ignored by BespokeASM and you could just as easily use a YAML comment), only the operand set `register_except_a` emits byte code . In your configuration [`integer`](https://github.com/sebseb7/logicpu/blob/master/isa.yaml#L25) and [`address`](https://github.com/sebseb7/logicpu/blob/master/isa.yaml#L196) only emit arguments, while the registers only emit byte code. In BespokeASM's model, these are two different things, and the instruction operand option `reverse_bytecode_order` only impacts byte code emitted. The whole idea is that byte code is the value that specifies the instruction, and the argument is used by the instruction's microcode one the instruction is loaded. In the current BespokeASM model, arguments always follow the instruction byte code after the byte code induced by the base instruction and all operands are constructed.

It sounds like you need to either turn the `integer` and `address` operands into byte code (you could use the `numeric_bytecode` operand type), or rethink your microcode some, or ask me to add that feature that `register` operand types can emit argument values instead of or in addition to byte code (though this would be a new paradigm because argument values are supposed to be interpreted from expression in the assembly code, a register argument would be more like an enumeration).

It might helpful to me to understand how your CPU's architecture works, as that might allow me to think of a better yet solution.

**Comment by sebseb7** [comment]
"var6" does the following:

it loads an address into the address register (therefore needs to be the first argument)
it then uses this address in the address register to load a byte to the low byte of the address register
then an integer is loaded in the high byte of the address register (because of this, the integer needs to come first)
then this address is loaded and put into a register 

because the address register is overwritten with data fetched using that same address register, the address for the low byte needs to come before the integer for the high byte.

by my convention it should be "ld <target_register>,<high_byte_of_addr_to_load_as_integer>,<low_byte_of_addr_to_load_as_addr>"

sim:
http://139.162.211.9:3003/

**Comment by sebseb7** [comment]
(never used yaml before and wasn't aware how comments work)

**Comment by sebseb7** [comment]
I just noticed that the target register can indeed come first in the argument list, because it's store in a differend register, but the integer and the addr needs to be in reversed order, because I cannot load the high byte of the address register first, because I need the same address register before that to load the low byte into the same address register.

**Comment by sebseb7** [comment]
probably not usefull to understand the architeture:

![image](https://user-images.githubusercontent.com/677956/179416419-d207b89a-a3e8-48f5-b9ad-6ef7581839ec.png)

the "intstruction register 2" stores the target register.

**Comment by michaelkamprath** [comment]
> I just noticed that the target register can indeed come first in the argument list, because it's stored in a differend register, but the integer and the addr needs to be in reversed order, because I cannot load the high byte of the address register first, because I need the same address register before that to load the low byte into the same address register.

So, are you sayin having the first operand as the target register emit byte code, and the second and third operand emit argument values that then get reversed between them does work for you? I am still not entirely clear as to what you want to happen. That is, what is the machine code that you want generated for the `ld dd,0x81,[var4]` instruction?

Keep in mind "arguments" are intended to be additional data in the machine code placed after an instruction byte code the the microcode loads through memory read operations as needed, and byte code is used to select the instruction being executed. I realize that things can get fuzzy between the two concepts, but just keep in mind BespokeASM treats instruction byte code and argument values as different things. Its interesting that you are having the microcode directly interpret an address value. that is, the address value is used to select microcode in in the microcode ROM rather than it be placed in something like a memory address register. Am I understanding that right?

**Comment by sebseb7** [comment]
probably Iam using the concept not as it was intended, have to check the documentation to really undersand it. but as the correct bytes are emitted so far in all other cases I never realized there is a difference between bytecode and argument values.

the machine code I want to generate for ```ld dd,0x81,[var4]``` is as if it was written ```ld [var4],0x81,dd``` , but I realized ```ld dd,[var4],0x81``` also works. I have no temporary register to store the 0x81 but I need to access it AFTER I accessed [var4]. When the programm counter moves over 0x81 I have no place to put it.

"that is, the address value is used to select microcode in in the microcode ROM rather than it be placed in something like a memory address register. Am I understanding that right?"

no. the address register needs to be used twice for this instruction, first to fetch the low byte of some address (using a 16 bit address) and then to fetch another value, using an address constructed from the low byte returned and the other argument (0x81), which is a high byte.

pseudocode:
byte x = fetch_address(var4) ; var4 is 16 bit, an address
register = fetch_address((0x81<<8)+x) ; also a 16 bit address

**Comment by michaelkamprath** [comment]
Ah, I see what you are trying to do.

You separately asked for the ability to do a relative jump. To me that implies your CPU has the ability to add a value to the program counter or more generally add an offset to the value on the address bus. Can you define step in microcode to add an 8-byte value to an address in the address register? (sorry, I am not deeply familiar with the notation in the logic diagram you shared, but I see the `sp_plus_16` operand type) If so, might I suggest another approach. The instruction syntax would be:

```asm
time:
    .byte 0
loop:
    add [time],1
    ld2 ar,sine       ; don't know if you have a 2-byte load variant, but that is what is intended here. The can be placed before loop too.
    ld a,[ar+[time]]
    jmp loop

.org 0x8100
sine:
    ....
```

The second operand in ` ld a,[ar+[time]]` is a `indirect_indexed_register` operand type. It generates byte code based on the register user (the `ar` in this example) and the source fo the offset value (the indirect value `[time]`), and has one argument her of the address used in the indirect value offset (`time`).  Your microcode would be different than what you explain, but I do feel this approach would be more clear than trying to fetch an indirect value at an offset from a base address then way you are. You can see an example of me doing the same thing on my CPU [here](https://github.com/michaelkamprath/eater-sap-1-improvements/blob/main/advanced-alu/examples/stringlib.asm#L227). 

Another approach you could use to get the syntax you are trying to make work is to use you stack for temporary values. You may not have a temporary register, but you should be able to in the scope and a single instruction's microcode push and pop a value onto the stack without disturbing stack for anything else. I don't know how many steps your microcode has, though. I use stack for temporary value storage in microcode often, such as [this example](https://github.com/michaelkamprath/eater-sap-1-improvements/blob/main/advanced-alu/microcode/putey-beta4-microcode.yaml#L8522-L8554) (it is `tstb immediate, bit_idx` instruction that will test a specific bit of an immediate value). 

Finally, for the pseudo-microcode you show above, could this line:
```
pc_to_addr_bus ; pc++ ; mem_out ; register_selector_in
```
be moved to be the first step in the microcode? If so, then the syntax you want can work if you just set `reverse_argument_order` for the instruction, as the only two operands with arguments are the immediate and indirect address, and the register operand only emits byte code which BespokeASM always puts before argument values.

**Comment by sebseb7** [comment]
currently the offset adders are hardwired to the stack pointer. there is a bus driver directly connecting the stack pointer to the address bus, and another bus driver connecting an adder to the address bus, which gets it's inputs from the stack pointer and the address register. it's a 16-bit adder, but only 8 bits are connected to the address-register. to use the offset only the low byte of the AR is used, because I rarely need large stack offsets, stack offsets are for local variables in subroutines, if a subroutine has more than 256 local variables, you probably should split it into more smaller subroutines.

![image](https://user-images.githubusercontent.com/677956/179483699-d19e1f82-d2ac-430b-9613-b446a5c77d82.png)


"the logic diagram you shared"

it's https://github.com/kevinawalsh/logisim-evolution , which greatly helped me to investigate differend designs before cutting thousands of wires. my design goal is to achive high clock-efficiency to 74xx part count ratio, while restricting myself to simple and easy to source chips (no programmable logic arrays or complex ALU chips) and to provide the minium capabilities to support a useful subset of C

```C
unsigned int abc(unsigned int a) { // a and the ret val both use the same stack position
   unsigned int b = 0; // lives on the stack, access using 8 bit sp offset
}
```

" don't know if you have a 2-byte load variant"

the AR register is invisible to assembly, only microcode can use it. this avoids the need to save it's content during interrupts. the AR register can only output to the address bus, not to the data bus.

```
    ld2 ar,sine 
    ld a,[ar+[time]]
```

also this requires more cycles than ```ld a,BYTE1(sine),[time]``` , the ld2 approach loads the low byte of sine, which is 0 anyway , because I align it at 0x??00, the video ram for the matrix display is also 0x??00 aligned, another heavily used application of this addressing mode.

"could this line: be moved to be the first step in the microcode?"

yes, but I noticed this only after I filled this "issue". so basically this specific issue is no longer open for me, but maybe it would be usefull to have total flexibility for ordering everything emitted. maybe this will pop up once more when I complete the instruction set. To achive high clock efficiency I will have a total of 32 variants of ```ld``` a future C compiler can use to optimize for speed. for example: the A register is a "fast" register, similar to the zero page of the 6502, because all opcodes using registers have a variant which ommits the register from the argument list, implying register A, registers B-H needs to be passed as an argument. ```ld a,1``` is 2 bytes opcode ```ld b,1``` is 3 bytes, this is why I have the registers_except_a operand set. same for the display data register, I use a 128x64 matrix display (128064B) which has the same hardware interface like the LCD2004, and Iam accessing it like a register.
